src/datachain/lib/dc/storage.py

```diff
@@ -79,40 +154,35 @@ def from_storage(
     cache = session.catalog.cache
     client_config = session.catalog.client_config
 
-    list_ds_name, list_uri, list_path, list_ds_exists = get_listing(
-        uri, session, update=update
-    )
+    uris = uri if isinstance(uri, (tuple, list)) else [uri]
 
-    # ds_name is None if object is a file, we don't want to use cache
-    # or do listing in that case - just read that single object
-    if not list_ds_name:
-        dc = from_values(
+    if not uris:
+        raise ValueError("No URIs provided")
+
+    dc = None
+    listed_uris = set()
+
+    for single_uri in uris:
+        update_for_uri = update
+        if update:
+            _, lst_uri, _ = parse_listing_uri(os.fsdecode(single_uri), client_config)
+            if lst_uri in listed_uris:
+                listed_uris.add(lst_uri)
+                update_for_uri = False
+
+        chain = _extract_chain(
+            uri=single_uri,
+            file_type=file_type,
+            cache=cache,
             session=session,
             settings=settings,
             in_memory=in_memory,
-            file=[get_file_info(list_uri, cache, client_config=client_config)],
+            recursive=recursive,
+            object_name=object_name,
+            update=update_for_uri,
+            client_config=client_config,
         )
-        dc.signals_schema = dc.signals_schema.mutate({f"{object_name}": file_type})
-        return dc
 
-    if update or not list_ds_exists:
-        # disable prefetch for listing, as it pre-downloads all files
-        (
-            from_records(
-                DataChain.DEFAULT_FILE_RECORD,
-                session=session,
-                settings=settings,
-                in_memory=in_memory,
-            )
-            .settings(prefetch=0)
-            .gen(
-                list_bucket(list_uri, cache, client_config=client_config),
-                output={f"{object_name}": File},
-            )
-            .save(list_ds_name, listing=True)
-        )
-
-    dc = from_dataset(list_ds_name, session=session, settings=settings)
-    dc.signals_schema = dc.signals_schema.mutate({f"{object_name}": file_type})
+        dc = dc.union(chain) if dc else chain
```

Comment:
can we avoid doing union for files? it can be quite expensive creating a table per each file in the list and then doing all these unions

it is indeed (most likely) needed for non-file sources (need a listing and then union them), unfortunately

Reply:
We have special case for files where we use dataset from values in case of file 

https://github.com/iterative/datachain/blob/main/src/datachain/lib/dc/storage.py#L84-L98

Reply:
will it be a single table per each file + union at the end?

Reply:
> will it be a single table per each file + union at the end?

Sorry, I don't understand, WDYM?

Reply:
does it call a single `from_records` per each single file?

Reply:
As you can see from the code linked above, it doesn't even call from records but call from_values for a single file

Reply:
okay, but it does it for every single file, right? essentially creating as many table in DB as the number of files you passed?

Reply:
Yeah, I think so. I will look into the possible way of creating a single table for each files passed,

Reply:
Updated PTAL